lumicks/pylake/kymotracker/detail/denoising.py

```diff
@@ -0,0 +1,412 @@
+           (Vol. 6, pp. VI-233). IEEE.
+    .. [2] Zhang, B., Fadili, J. M., & Starck, J. L. (2008). Wavelets, ridgelets, and curvelets
+           for Poisson noise removal. IEEE Transactions on image processing, 17(7), 1093-1108.
+    """
+
+    def __init__(self, kernels, *, two_dimensional=False):
+        self._kernels = [k * k.T for k in kernels] if two_dimensional else kernels
+
+        # Full kernels contain h0 * h1 * ... * hn where * is convolution.
+        equal_length_kernels = equal_length(generate_product_filters(kernels))
+        self._full_kernels = [k * k.T if two_dimensional else k for k in equal_length_kernels]
+        self._coefficients = [calculate_vst_coefficients(k) for k in self._full_kernels]
+
+        self._stdev = [
+            calculate_vst_stdev(k1, k2)
+            for k1, k2 in zip(self._full_kernels[1:], self._full_kernels[:-1])
+        ]
+
+    def _calculate_wavelets(self, image, stabilize):
+        """Calculate the wavelet coefficient images and remainder of a IUWT wavelet transform
+
+        Parameters
+        ----------
+        image : np.ndarray
+            Image to decompose into wavelet coefficients.
+        stabilize : bool
+            Applies a transformation that stabilizes the variance for the coefficient images.
+
+        Returns
+        -------
+        detail_coefficients : List[np.ndarray]
+            List of images representing the detail coefficients for each wavelet.
+        remainder : np.ndarray
+            Image containing the remainder (what's left after adding the detail layers).
+        """
+        filtered_imgs = [convolve(image, kernel) for kernel in self._full_kernels]
+
+        # Stabilize variance
+        if stabilize:
+            filtered_imgs = [
+                variance_stabilizing_transform(img, coeffs)
+                for img, coeffs in zip(filtered_imgs, self._coefficients)
+            ]
+
+        detail_coefficients = [f1 - f2 for f1, f2 in zip(filtered_imgs, filtered_imgs[1:])]
+        remainder = filtered_imgs[-1]
+
+        return detail_coefficients, remainder
+
+    def _reconstruct_image(self, detail_coefficients, remainder, stabilize):
+        """Reconstructs image from wavelet decomposition.
+
+        Parameters
+        ----------
+        detail_coefficients : List[np.ndarray]
+            List of images representing the detail coefficients for each wavelet.
+        remainder : np.ndarray
+            Image containing the remainder (what's left after adding the detail layers).
+        stabilize : bool
+            Indicate that the detail coefficients were stabilized and this transform
+            must be inverted.
+        """
+        output_image = sum(detail_coefficients + [remainder])
+        if not stabilize:
+            return output_image
+
+        return inverse_variance_stabilizing_transform(output_image, self._coefficients[0])
+
+    def process_image(self, image, false_detection_rate=0.1, verbose=False):
```

Comment:
`filter_image()`? In general I don't really like using "process", it's pretty vague sounding